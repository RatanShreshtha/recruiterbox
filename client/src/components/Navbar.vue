<template>
  <b-navbar toggleable="lg" type="light" variant="light">
    <b-navbar-brand href="/">Recruiterbox Kudos</b-navbar-brand>

    <b-navbar-toggle target="nav-collapse"></b-navbar-toggle>

    <b-collapse id="nav-collapse" is-nav>
      <b-navbar-nav v-if="token">
        <b-nav-item id="kudos-received" href="/kudos-received"
          >Kudos Received</b-nav-item
        >
        <b-nav-item id="kudos-given" href="/kudos-given"
          >Kudos Given</b-nav-item
        >
      </b-navbar-nav>

      <b-navbar-nav class="ml-auto">
        <b-navbar-nav v-if="token">
          <b-nav-item href="/">
            {{ first_name }} {{ last_name }} ({{ username }})
          </b-nav-item>
          <b-nav-item href="/kudo-add" v-if="kudos > 0">
            <b-button size="sm" type="button" variant="outline-info">
              Give Kudo
            </b-button>
          </b-nav-item>
          <b-nav-item href="#" class="mt-auto">
            <b-button
              id="logout"
              size="sm"
              type="button"
              variant="outline-danger"
              @click="processLogout"
            >
              Log Out
            </b-button>
          </b-nav-item>
        </b-navbar-nav>
      </b-navbar-nav>
    </b-collapse>
  </b-navbar>
</template>

<script>
import { mapState, mapActions } from "vuex";

export default {
  name: "Navbar",
  computed: {
    ...mapState(["username", "first_name", "last_name", "token", "kudos"])
  },
  methods: {
    ...mapActions(["logout"]),
    async processLogout() {
      this.logout();
    }
  }
};
</script>

<style scoped></style>
