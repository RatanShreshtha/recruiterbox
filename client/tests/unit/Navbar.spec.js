import { shallowMount, createLocalVue } from "@vue/test-utils";

import { createRenderer } from "vue-server-renderer";

import Vuex from "vuex";
import VueRouter from "vue-router";
import { BootstrapVue, IconsPlugin } from "bootstrap-vue";

import Navbar from "@/components/Navbar.vue";

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueRouter);

localVue.use(BootstrapVue);
localVue.use(IconsPlugin);

describe("Navbar.vue", () => {
  describe("User is not Logged In", () => {
    let state;
    let store;

    beforeEach(() => {
      state = {
        id: null,
        kudos: 0,
        username: null,
        first_name: null,
        last_name: null,
        email: null,
        token: null
      };

      store = new Vuex.Store({
        state
      });
    });

    it("matches snapshot", () => {
      const renderer = createRenderer();

      const wrapper = shallowMount(Navbar, {
        store,
        localVue
      });

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err);
        expect(str).toMatchSnapshot();
      });
    });
  });

  describe("User is Logged In", () => {
    let actions;
    let router;
    let routes;
    let state;
    let store;

    beforeEach(() => {
      state = {
        id: 1,
        kudos: 3,
        username: "root",
        first_name: "I am",
        last_name: "Root",
        email: "iam.root@example.com",
        token: "cd6ee1b3f55ddf30457276d6d9a9c17ca03f37b0"
      };

      actions = {
        logout: jest.fn()
      };

      routes = [
        {
          path: "/",
          name: "Home"
        }
      ];

      router = new VueRouter({
        routes
      });

      store = new Vuex.Store({
        state,
        actions
      });
    });

    it("matches snapshot", () => {
      const renderer = createRenderer();

      const wrapper = shallowMount(Navbar, {
        router,
        store,
        localVue
      });

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err);
        expect(str).toMatchSnapshot();
      });
    });

    it("log out user when clicked on log out", () => {
      const wrapper = shallowMount(Navbar, {
        router,
        store,
        localVue
      });

      expect(wrapper.vm.$route.path).toBe("/");

      const button = wrapper.find("#logout");

      button.trigger("click");

      expect(wrapper.vm.$route.path).toBe("/");
    });
  });
});
