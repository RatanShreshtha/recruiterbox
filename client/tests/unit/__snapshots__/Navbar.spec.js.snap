// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar.vue User is Logged In matches snapshot 1`] = `
<b-navbar-stub tag="nav" type="light" variant="light" toggleable="lg">
  <b-navbar-brand-stub href="/" target="_self" event="click" routerTag="a" tag="div">Recruiterbox Kudos</b-navbar-brand-stub>
  <b-navbar-toggle-stub label="Toggle navigation" target="nav-collapse"></b-navbar-toggle-stub>
  <b-collapse-stub id="nav-collapse" isNav="true" tag="div">
    <b-navbar-nav-stub tag="ul">
      <b-nav-item-stub href="/kudos-received" target="_self" event="click" routerTag="a" id="kudos-received">Kudos Received</b-nav-item-stub>
      <b-nav-item-stub href="/kudos-given" target="_self" event="click" routerTag="a" id="kudos-given">Kudos Given</b-nav-item-stub>
    </b-navbar-nav-stub>
    <b-navbar-nav-stub tag="ul" class="ml-auto">
      <b-navbar-nav-stub tag="ul">
        <b-nav-item-stub href="/" target="_self" event="click" routerTag="a">
          I am Root (root)
        </b-nav-item-stub>
        <b-nav-item-stub href="/kudo-add" target="_self" event="click" routerTag="a">
          <b-button-stub target="_self" event="click" routerTag="a" size="sm" variant="outline-info" type="button" tag="button">
            Give Kudo
          </b-button-stub>
        </b-nav-item-stub>
        <b-nav-item-stub href="#" target="_self" event="click" routerTag="a" class="mt-auto">
          <b-button-stub target="_self" event="click" routerTag="a" size="sm" variant="outline-danger" type="button" tag="button" id="logout">
            Log Out
          </b-button-stub>
        </b-nav-item-stub>
      </b-navbar-nav-stub>
    </b-navbar-nav-stub>
  </b-collapse-stub>
</b-navbar-stub>
`;

exports[`Navbar.vue User is not Logged In matches snapshot 1`] = `
<b-navbar-stub tag="nav" type="light" variant="light" toggleable="lg">
  <b-navbar-brand-stub href="/" target="_self" event="click" routerTag="a" tag="div">Recruiterbox Kudos</b-navbar-brand-stub>
  <b-navbar-toggle-stub label="Toggle navigation" target="nav-collapse"></b-navbar-toggle-stub>
  <b-collapse-stub id="nav-collapse" isNav="true" tag="div">
    <!---->
    <b-navbar-nav-stub tag="ul" class="ml-auto">
      <!---->
    </b-navbar-nav-stub>
  </b-collapse-stub>
</b-navbar-stub>
`;
